keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,432 @@
+
+#### Deploy Windows workloads that consume persistent storage managed by a CSI plugin
+
+An operator should be able to deploy a Windows workload like SQL Server that consumes dynamically provisioned Persistent Volumes managed by a CSI plugin using:
+
+A storage class like:
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gce-pd
+provisioner: pd.csi.storage.gke.io
+parameters:
+  type: pd-standard
+```
+
+with a PVC like:
+```
+apiVersion: v1
+metadata:
+  name: sqlpvc
+spec:
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: csi-gce-pd
+  resources:
+    requests:
+      storage: 100Gi
+```
+
+and a Pod like:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: sqlserver1
+spec:
+  tolerations:
+  - key: "node.kubernetes.io/os"
+    operator: "Equal"
+    value: "win1809"
+    effect: "NoSchedule"
+  nodeSelector:
+    beta.kubernetes.io/os: windows
+  containers:
+  - name: sqlpod
+    image: ddebroy/sqlsrv:latest
+    volumeMounts:
+    - name: csi-sql-vol
+      mountPath: C:\Data
+    env:
+    - name: ACCEPT_EULA
+      value: "y"
+    - name: sa_password
+      value: "*****"
+  volumes:
+    - name: csi-sql-vol
+      persistentVolumeClaim:
+        claimName: sqlpvc
+```
+
+### Implementation Details
+
+CSI Node Plugins listen on domain sockets and respond to CSI API requests sent over GRPC from a container orchestrator like Kubernetes. They are responsible all storage management operations scoped around a specific node that are typically necessary after a CSI Controller Plugin has finished provisioning a Persistent Volume and attached it to the node. In Kubernetes, the CSI Node API calls are invoked by the CSI In-tree Plugin in the kubelet as well as the CSI Node Driver Registrar. The CSI Node Driver Registrar interacts with the Kubelet Plugin Watcher and it is maintained by the Kubernetes CSI community as a side-car container for deployment in CSI Node Plugin pods.
+
+![Kubernetes CSI Components](https://raw.githubusercontent.com/kubernetes/community/master/contributors/design-proposals/storage/container-storage-interface_diagram1.png?raw=true "Kubernetes CSI Components")
+
+Support for Unix Domain Sockets has been introduced in Windows Server 2019 and works across containers as well as host and container as long as the processes running in containers are listening on the socket. If a process from within a container wishes to connect to a domain socket that a process on the host OS is listening on, Windows returns a permission error. This scenario however does not arise in the context of interactions between Kubelet, CSI Node Driver Registrar and CSI Node Plugin as these involve a process in a container listening on a domain socket (CSI Node Driver Registrar or CSI Node Plugin) that a process on the host (Kubelet) connects to.
```

Comment:
What about the proxy? Host listening and the process in the container connecting. Also is there documentation or reasoning on why this restriction exists?

Reply:
The proxy exposes a Windows named pipe. Will call it out here and mention it is detailed in later sections. Regarding why this restriction exists, given the very specific permission error when trying to connect to a host socket from a containerized process, my assumption is it's a conscious security choice at the OS layer. Unfortunately, this behavior or error is not clearly documented in MSDN.

Reply:
@liggitt Windows named pipes are an established way to have a container communicate with a host process for a while now in Windows. Support was added back in 2016: https://github.com/moby/moby/pull/19911 in Docker engine and was specifically intended to provide a mechanism similar to Unix Domain Sockets in Windows before support for Domain Sockets was enabled in Windows later in 2019. /cc @PatrickLang for any additional comments.

Reply:
Honestly I'm not sure why it's being blocked. It may be a regression in Windows and it's something we can root cause as we implement this. Named pipes and unix domain sockets are functionally equivalent and both work as container mounts, so I think we'll decide on which to enable as part of this work. We've already shown that named pipes work, so if unix domain sockets can be "fixed" we can choose to enable that later.